File: android/Ficheros/Ficheros/app/src/main/java/com/mastermoviles/persistencia/files/MainActivity.java
package com.mastermoviles.persistencia.files;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.OutputStreamWriter;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private Button externalState;
    private EditText editText;
    private Button addToFile;
    private Button showFile;
    private Button moveExternal;
    private Button moveInternal;
    private Button finishApp;

    private final static int REQUEST_PERMISSIONS_CODE = 1;
    private static final String[] PERMISSIONS = new String[] {
            Manifest.permission.WRITE_EXTERNAL_STORAGE,
            Manifest.permission.READ_EXTERNAL_STORAGE
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        this.setUIContent();
        this.checkPermissions();
    }

    private void setUIContent() {
        externalState = findViewById(R.id.externalState);
        editText = findViewById(R.id.editText);
        addToFile = findViewById(R.id.addToFile);
        showFile = findViewById(R.id.showFile);
        moveExternal = findViewById(R.id.moveExternal);
        moveInternal = findViewById(R.id.moveInternal);
        finishApp = findViewById(R.id.finish);

        externalState.setOnClickListener(this);
        addToFile.setOnClickListener(this);
        showFile.setOnClickListener(this);
        moveExternal.setOnClickListener(this);
        moveInternal.setOnClickListener(this);
        finishApp.setOnClickListener(this);
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.externalState:
                Intent intent = new Intent(this, ExternalActivity.class);
                startActivity(intent);
                break;
            case R.id.addToFile:
                this.saveDataToFile();
                break;
            case R.id.showFile:
                Intent showFileIntent = new Intent(this, FileActivity.class);
                startActivity(showFileIntent);
                break;
            case R.id.moveExternal:
                this.moveToExternal();
                break;
            case R.id.moveInternal:
                this.moveToInternal();
                break;
            case R.id.finish:
                finish();
                break;
            default:
                Log.d("Main", "no existe el botón");
        }
    }

    private void saveDataToFile() {
        String text = editText.getText().toString();
        try {
            OutputStreamWriter fout = new OutputStreamWriter(openFileOutput("practica_ficheros.txt", Context.MODE_APPEND));
            fout.append(text + "\n");
            fout.close();
            Toast.makeText(this, "Texto guardado con éxito", Toast.LENGTH_LONG).show();
            editText.setText("");
        }
        catch (Exception ex) {
            Log.e("Files", String.format("Error: %s", ex.getMessage()));
        }
    }

    private void moveToExternal() {
        if (checkSDCardStatus()) {
            File path = null;
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.KITKAT) {
                path = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS);
            }
            else {
                path = new File(Environment.getExternalStorageDirectory().getAbsolutePath());
            }
            File file = new File(path, "practica_ficheros.txt");
            try {
                path.mkdirs();
                InputStream inputStream = openFileInput("practica_ficheros.txt");
                OutputStream outputStream = new FileOutputStream(file);
                byte [] data = new byte[inputStream.available()];
                inputStream.read(data);
                outputStream.write(data);
                inputStream.close();
                outputStream.close();
                Toast.makeText(this, "Se ha movido el fichero a la memoria externa", Toast.LENGTH_LONG).show();
            }
            catch (IOException ex) {
                Log.w("ExternalStorage", "Error writting: " + file, ex);
                String message = "Se ha producido un error al escribir en la memoria SD";
                if (ex.getMessage().contains("ENOENT")) {
                    message = "No existe el fichero que se quiere mover.";
                }
                Toast.makeText(this, message, Toast.LENGTH_LONG).show();
            }
        }
        else {
            Toast.makeText(this, "No se puede acceder a la memoroia SD", Toast.LENGTH_LONG).show();
        }
    }

    private void moveToInternal() {
        if (checkSDCardStatus()) {
            File sdPath = null;
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                sdPath = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS);
            }
            else {
                sdPath = new File(Environment.getExternalStorageDirectory().getAbsolutePath());
            }
            File file = new File(sdPath, "practica_ficheros.txt");

            try {
                InputStream inputStream = new FileInputStream(file);
                OutputStream outputStream = openFileOutput("practica_ficheros.txt", Context.MODE_PRIVATE);
                byte [] data = new byte[inputStream.available()];
                inputStream.read(data);
                outputStream.write(data);
                inputStream.close();
                outputStream.close();
                Toast.makeText(this, "Se ha movido el fichero a la memoria interna", Toast.LENGTH_LONG).show();
            }
            catch (IOException ex) {
                Log.w("ExternalStorage", "Error writting: " + file, ex);
                String message = "Se ha producido un error al escribir en la memoria interna";
                if (ex.getMessage().contains("ENOENT")) {
                    message = "No existe el fichero que se quiere mover.";
                }
                Toast.makeText(this, message, Toast.LENGTH_LONG).show();
            }
        }
        else {
            Toast.makeText(this, "No se puede acceder a la memoroia SD", Toast.LENGTH_LONG).show();
        }
    }

    private boolean checkSDCardStatus() {
        if (Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED)) {
            return true;
        }
        else if (Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED_READ_ONLY)) {
            return false;
        }
        else {
            return false;
        }
    }

    private boolean checkPermissions() {
        if(Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            return true;
        }
        if(ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED &&
            ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
            return true;
        }
        else {
            requestPermissions(PERMISSIONS, REQUEST_PERMISSIONS_CODE);
        }
        return false;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_PERMISSIONS_CODE) {
            for(int i = 0; i < grantResults.length; i++) {
                if(grantResults[i] != PackageManager.PERMISSION_GRANTED) {
                    finish();
                    break;
                }
            }
            return;
        }
        else {
            finish();
        }
    }
}

File: android/SQLite/app/src/main/java/com/mastermoviles/persistencia/sqlite/MainActivity.kt
package com.mastermoviles.persistencia.sqlite

import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.mastermoviles.persistencia.sqlite.Data.Database
import com.mastermoviles.persistencia.sqlite.Models.Users
import com.mastermoviles.persistencia.sqlite.UserManagement.ManagementActivity
import com.mastermoviles.persistencia.sqlite.UserManagement.UserActivity
import java.io.*

class MainActivity : AppCompatActivity() {

    var userName: EditText? = null
    var password: EditText? = null
    var login: Button? = null
    var close: Button? = null
    val database = Database(this)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        this.setUIContent()
    }

    private fun setUIContent() {
        userName = findViewById(R.id.userName)
        password = findViewById(R.id.password)
        login = findViewById(R.id.login)
        close = findViewById(R.id.close)

        login?.setOnClickListener{
            this.login()
        }

        close?.setOnClickListener{
            finish()
        }
    }

    private fun login(): Boolean {
        val name = userName?.text.toString()
        val pass = password?.text.toString()
        if (name.isEmpty() || pass.isEmpty()) {
            Toast.makeText(
                this,
                "Debes introducir el nombre de usuario y la contraseña",
                Toast.LENGTH_SHORT
            ).show()
            return false
        }
        val user: Users? = database.login(name)
        return when {
            user == null -> {
                Toast.makeText(this, "No existe ningún usuario con ese nombre", Toast.LENGTH_SHORT).show()
                false
            }
            user.password != pass -> {
                Toast.makeText(this, "La contraseña no es correcta", Toast.LENGTH_SHORT).show()
                false
            }
            else -> {
                val intent = Intent(this, UserActivity::class.java)
                intent.putExtra("fullName", user.nombre_completo)
                intent.putExtra("userName", user.nombre_usuario)
                startActivity(intent)
                true
            }
        }
    }

    private fun checkExternalStatus(): Boolean {
        return if (Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED) {
            true
        } else if (Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED_READ_ONLY) {
            false
        } else {
            false
        }
    }

    private fun backup() {
        if (checkExternalStatus()) {
            val users = database.getUsers()
            try {
                val fileOutput = OutputStreamWriter(openFileOutput("practicaSQLite.txt", Context.MODE_PRIVATE))
                users.forEach { u->
                    fileOutput.append("${u.id}-${u.nombre_usuario}-${u.password}-${u.nombre_completo}-${u.email}\n")
                }
                fileOutput.close()
                Toast.makeText(this, "Backup creado con éxito", Toast.LENGTH_SHORT).show()
            }
            catch(ex: IOException) {
                ex.printStackTrace()
            }
        }
        else {
            Toast.makeText(this,"No se puede acceder al almacenamiento externo", Toast.LENGTH_SHORT).show()
        }
    }

    private fun restore() {
        if (checkExternalStatus()) {
            try {
                val fileInput = BufferedReader(InputStreamReader(openFileInput("practicaSQLite.txt")))
                var line = fileInput.readLine()
                while (line != null) {
                    val values = line.split("-")
                    Log.d("Main", "" + values[1])
                    val user = Users(null, values[1], values[2], values[3], values[4])
                    val isAdded = database.addUser(user)
                    line = fileInput.readLine()
                }
                fileInput.close()
                Toast.makeText(this, "Backup restaurado con éxito", Toast.LENGTH_SHORT).show()
            }
            catch(ex: IOException) {
                ex.printStackTrace()
                Toast.makeText(this, "No se puede abrir o no existe el backup", Toast.LENGTH_SHORT).show()
            }
        }
        else {
            Toast.makeText(this,"No se puede acceder al almacenamiento externo", Toast.LENGTH_SHORT).show()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.backup -> this.backup()
            R.id.restore -> this.restore()
            R.id.users -> {
                val intent = Intent(this, ManagementActivity::class.java)
                startActivity(intent)
            }
            else -> Toast.makeText(this, "No existe la opción seleccionada", Toast.LENGTH_SHORT).show()
        }
        return true
    }
}